src/collapse/perf.rs

```diff
@@ -412,3 +560,129 @@ fn tidy_java(mut func: String) -> String {
+                    "{}{}",
+                    &s[0..start],
+                    &s[start + disc.len() + end + 1..s.len()]
+                ));
+            }
+        }
+    }
+    s
+}
+
+// Removes Rust hash from function name if it exists.
+// When demangling Rust names, addr2line in binutils removes the hash,
+// but the Rust addr2line crate does not. Calling this function
+// after demangling will make it match binutils.
+fn strip_rust_hash(func: &str) -> &str {
+    if let Some(idx) = func.rfind("::") {
+        if is_rust_hash(&func[idx + 2..]) {
+            return &func[0..idx];
+        }
+    }
+    func
+}
+
+// Rust hashes are hex digits with an `h` prepended.
+fn is_rust_hash(s: &str) -> bool {
+    s.starts_with('h') && s[1..].chars().all(|c| c.is_digit(16))
+}
+
+// Returns whether the module can be un-inlined.
+//
+// Equivalent to the following perl:
+//     $mod !~ m/(perf-\d+.map|kernel\.|\[[^\]]+\])/;
+fn can_un_inline(module: &str) -> bool {
+    // /perf-\d+.map/
+    if let Some(perf) = module.find("perf-") {
+        let rest = &module[perf + 5..];
+        if let Some(map) = rest.find(".map") {
+            if rest[..map].chars().all(|c| c.is_ascii_digit()) {
+                return false;
+            }
+        }
+    }
+
+    // /kernel\./
+    if module.find("kernel.").is_some() {
+        return false;
+    }
+
+    // /\[[^\]]+\]/
+    let mut start = 0;
+    while start < module.len() {
+        if let Some(open) = module[start..].find('[') {
+            let rest = &module[open + 1..];
+            if let Some(close) = rest.find(']') {
+                if close > 1 && rest[..close].chars().all(|c| c != ']') {
+                    return false;
+                }
+            }
+            start = open + 1;
+        } else {
+            break;
+        }
+    }
+
+    true
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_remove_discriminator() {
+        assert_eq!(
+            remove_discriminator(Cow::from("123 (discriminator 456)")),
+            Cow::from("123")
+        );
+        assert_eq!(
+            remove_discriminator(Cow::from("foo:86 (discriminator 98765)")),
+            Cow::from("foo:86")
+        );
+    }
+
+    #[test]
+    fn test_strip_rust_hash() {
+        assert_eq!(
+            strip_rust_hash("inferno::collapse::perf::h123456789ABCDEF0"),
+            "inferno::collapse::perf"
+        );
+        assert_eq!(strip_rust_hash("foo::bar::h21F2A89C449"), "foo::bar");
+        assert_eq!(
+            strip_rust_hash("foo::bar::21F2A89C449"),
+            "foo::bar::21F2A89C449"
+        );
+        assert_eq!(strip_rust_hash("foo::bar"), "foo::bar");
+    }
+
+    #[test]
+    fn test_is_rust_hash() {
+        assert!(is_rust_hash("h21F2A89C449"));
+        assert!(!is_rust_hash("21F2A89C449"));
+        assert!(!is_rust_hash("q21F2A89C449"));
+        assert!(!is_rust_hash("h21F2A89C44Z"));
+    }
+
+    #[test]
+    fn test_can_un_iline() {
```

Comment:
I wish we had some real strings to use here that came from real profiles and that we know appear in the wild. That way we'd actually know what it's trying to match on...

Reply:
I'll use real test strings taken from the FlameGraph test files.